root/usr/share/fdi/facts/openlldp.rb

```diff
@@ -0,0 +1,99 @@
+# Fact:
+#   lldp_neighbor_chassisid_<interface>
+#   lldp_neighbor_mngaddr_ipv4_<interface>
+#   lldp_neighbor_mngaddr_ipv6_<interface>
+#   lldp_neighbor_mtu_<interface>
+#   lldp_neighbor_portid_<interface>
+#   lldp_neighbor_sysname_<interface>
+#   lldp_neighbor_pvid_<interface>
+#
+# Purpose:
+#   Return information about the host's LLDP neighbors.
+#
+# Resolution:
+#   On hosts with the lldptool binary, send queries to the lldpad for each of
+#   the host's Ethernet interfaces and parse the output.
+#
+# Caveats:
+#   Assumes that the connected Ethernet switch is sending LLDPDUs, Open-LLDP
+#   (lldpad) is running, and lldpad is configured to receive LLDPDUs on each
+#   Ethernet interface.
+#
+# Authors:
+#   Mike Arnold <mike@razorsedge.org>
+#
+# Copyright:
+#   Copyright (C) 2012 Mike Arnold, unless otherwise noted.
+#
+
+require 'facter/util/macaddress'
+
+# http://www.ruby-forum.com/topic/3418285#1040695
+module Enumerable
+  def grep_v(cond)
+    select {|x| not cond === x}
+  end
+end
+
+#if Facter::Util::Resolution.which('lldptool')
+if File.exists?('/usr/sbin/lldptool')
+  lldp = {
+    # LLDP Name    Numeric value
+    'chassisID'    => '1',
+    'portID'       => '2',
+    'sysName'      => '5',
+    'mngAddr_ipv4' => '8',
+    'mngAddr_ipv6' => '8',
+    'PVID'         => '0x0080c201',
+    'MTU'          => '0x00120f04',
+  }
+
+  # Remove interfaces that pollute the list (like lo and bond0).
+  Facter.value('interfaces').split(/,/).grep_v(/^lo$|^bond[0-9]/).each do |interface|
+    # Loop through the list of LLDP TLVs that we want to present as facts.
+    lldp.each_pair do |key, value|
+      Facter.add("lldp_neighbor_#{key}_#{interface}") do
+        setcode do
+          result = ""
+          output = Facter::Util::Resolution.exec("lldptool get-tlv -n -i #{interface} -V #{value} 2>/dev/null")
```

Comment:
nitpick: I would extract the exec from both the lldp.each_pair and facter.add loops, it can be the first thing you do, and not enter any other code if there is no output.